include/nbl/video/IDeviceMemoryBacked.h

```diff
@@ -1,44 +1,75 @@
-// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// Copyright (C) 2018-2022 - DevSH Graphics Programming Sp. z O.O.
 // This file is part of the "Nabla Engine".
 // For conditions of distribution and use, see copyright notice in nabla.h
-
-#ifndef __NBL_I_DRIVER_MEMORY_BACKED_H_INCLUDED__
-#define __NBL_I_DRIVER_MEMORY_BACKED_H_INCLUDED__
+#ifndef _NBL_I_DRIVER_MEMORY_BACKED_H_INCLUDED_
+#define _NBL_I_DRIVER_MEMORY_BACKED_H_INCLUDED_
 
 #include "IDeviceMemoryAllocation.h"
 #include <algorithm>
 
-#define VK_NO_PROTOTYPES
-#include <vulkan/vulkan.h>
-
 namespace nbl::video
 {
 
-//! Interface from which resources backed by IDeviceMemoryAllocation, such as ITexture and IGPUBuffer, inherit from
+//! Interface from which resources backed by IDeviceMemoryAllocation inherit from
 class IDeviceMemoryBacked : public virtual core::IReferenceCounted
 {
     public:
+        //! If you bound an "exotic" memory object to the resource, you might require "special" cleanups in the destructor
+        struct ICleanup
+        {
+            virtual ~ICleanup() = 0;
+        };
+
+        //!
+        struct SCachedCreationParams
+        {
+            // A Pre-Destroy-Step is called out just before a `vkDestory` or `glDelete`, this is only useful for "imported" resources
+            std::unique_ptr<ICleanup> preDestroyCleanup = nullptr;
+            // A Post-Destroy-Step is called in this class' destructor, this is only useful for "imported" resources
+            std::unique_ptr<ICleanup> postDestroytCleanup = nullptr;
+            // If non zero, then we're doing concurrent resource sharing
+            uint8_t queueFamilyIndexCount = 0u;
+            // Whether the handle is imported, thus the destructor will skip the call to `vkDestroy` or `glDelete` on the handle, this is only useful for "imported" objects
+            bool importedHandle = false;
```

Comment:
@Erfan-Ahmadi just because you've imported a handle doesn't mean you should own it (be in charge of deletion) please pick a different name

Reply:
Your comment doesn't make sense to me. Do you mean just because I imported a handle doesn't mean I should skip the deletion?

Reply:
> Your comment doesn't make sense to me. Do you mean just because I imported a handle doesn't mean I should skip the deletion?

yes.

Skipping deletion implies the handle was imported, but importing a handle does not imply skipping deletion.